crates/client/mempool/src/inner/nonce_chain.rs

```diff
@@ -0,0 +1,147 @@
+use super::tx::{ArrivedAtTimestamp, MempoolTransaction};
+use crate::TxInsersionError;
+use starknet_api::{core::Nonce, transaction::TransactionHash};
+use std::collections::{btree_map, BTreeMap};
+use std::{cmp, iter};
+
+#[derive(Debug, Clone)]
+pub struct OrderMempoolTransactionByNonce(pub MempoolTransaction);
+
+impl PartialEq for OrderMempoolTransactionByNonce {
+    fn eq(&self, other: &Self) -> bool {
+        self.cmp(other).is_eq()
+    }
+}
+impl Eq for OrderMempoolTransactionByNonce {}
```

Comment:
we only want the nonce to be compared, not the rest.
if two txs are equal, we want the two mempool transactions to appear as equal to the btreemap, so that we find that the entry is not empty for this nonce - this is how we check for nonce collision.
this goes for partialord/ord/partial_eq/eq, the OrderMempoolTransactionByNonce is a wrapper around MempoolTransaction that just orders them in the specific order so that the NonceChain::transactions btreemap works like we want it to